src/eth-query.ts

```diff
@@ -0,0 +1,656 @@
+
+  /**
+   * Makes an `eth_getFilterChanges` request.
+   *
+   * @param args - The params for the request followed by a callback, which will
+   * be called with an error value or a result value depending on the outcome of
+   * the request.
+   */
+  getFilterChanges = generateFnFor('eth_getFilterChanges');
+
+  /**
+   * Makes an `eth_getFilterLogs` request.
+   *
+   * @param args - The params for the request followed by a callback, which will
+   * be called with an error value or a result value depending on the outcome of
+   * the request.
+   */
+  getFilterLogs = generateFnFor('eth_getFilterLogs');
+
+  /**
+   * Makes an `eth_getLogs` request.
+   *
+   * @param args - The params for the request followed by a callback, which will
+   * be called with an error value or a result value depending on the outcome of
+   * the request.
+   */
+  getLogs = generateFnFor('eth_getLogs');
+
+  /**
+   * Makes an `eth_getWork` request.
+   *
+   * @param args - The params for the request followed by a callback, which will
+   * be called with an error value or a result value depending on the outcome of
+   * the request.
+   */
+  getWork = generateFnFor('eth_getWork');
+
+  /**
+   * Makes an `eth_submitWork` request.
+   *
+   * @param args - The params for the request followed by a callback, which will
+   * be called with an error value or a result value depending on the outcome of
+   * the request.
+   */
+  submitWork = generateFnFor('eth_submitWork');
+
+  /**
+   * Makes an `eth_submitHashrate` request.
+   *
+   * @param args - The params for the request followed by a callback, which will
+   * be called with an error value or a result value depending on the outcome of
+   * the request.
+   */
+  submitHashrate = generateFnFor('eth_submitHashrate');
+
+  /**
+   * Makes a request for a particular RPC method.
+   *
+   * @param opts - The JSON-RPC payload. Must include `method`, but may include
+   * `id`, `jsonrpc`, or `params`; reasonable defaults will be used for any
+   * omitted properties.
+   * @param callback - A function which will be called with an error value or a
+   * result value depending on the outcome of the request.
+   */
+  sendAsync<Params extends BaseParams, Result>(
+    opts: Partial<SendAsyncPayload<Params>> &
+      Pick<SendAsyncPayload<Params>, 'method'>,
+    callback: SendAsyncCallback<Result>,
+  ) {
+    const payload = createPayload(opts);
+    this.currentProvider.sendAsync<Params, Result>(
+      payload,
+      (error, response) => {
+        let improvedError = error;
+        if (!error && response.error) {
+          improvedError = new Error(
+            `EthQuery - RPC Error - ${response.error.message}`,
+          );
+        }
+        if (improvedError) {
+          return callback(improvedError, undefined);
+        }
+        if (response.result) {
+          return callback(null, response.result);
+        }
+        throw new Error(
```

Comment:
This error does not get thrown in the original implementation, but I added it because of the type that we are assigning to the response:

``` typescript
type ProviderSendAsyncResponse<Result> = {
  error?: { message: string };
  result?: Result;
};
```

In other words, both `error` and `result` can be `undefined`.

Of course we control the type here. That said, I derived it from the implementation. It would be a little more accurate to use:

``` typescript
type ProviderSendAsyncResponse<Result> =
  | {
      error: { message: string }
    }
  | {
      result: Result
    };
```

Or even (matching `@metamask/eth-json-rpc-provider`):

``` typescript
type ProviderSendAsyncResponse<Result> =
  | {
      error: unknown
    }
  | {
      result: Result
    };
```

but then we'd have to change the implementation.